### DISCUSSION TITLE
While loop not exiting

### DISCUSSION THREAD

**Discussion by MrTinkerman** [micropython/micropython#10459] [Using MicroPython]
Hi all,

I have a fairly simple program to start and stop a motor, see below. 
-- the motor is started and runs until both end stops go low (yes I have two end stops on the same side)
-- then it pauses and reverses

It seems to be working just fine but something stange is going on in this while loop:

```
while endstop_1.value() or endstop_2.value():
        
        print(endstop_1.value(), endstop_2.value(), time.ticks_ms())
        time.sleep_ms(20)
        led.value(1)
```
If I put in time.sleep_ms(20) :
-- the values for the end stops are printed out correctly, although with a slight delay
-- the led is on while the loop runs and is turned off after the motor sequence
-- when triggering both end stops the loop exits and the motor stops and reverses

However, if I comment out time.sleep_ms(20) :
-- the values for the end stops keep being printed as 1 (triggering pulls them low)
-- the led behavior is the same
-- when triggering both end stops the loop **seems to exit** and the motor stops and reverses

And here's the thing: the print statement keeps running ?? Even while it seems the loop worked just fine.
I keep getting '1 1 ticks_ms' printed out indefinitely. The led is not turned on again so you would say the code in the loop is not running.
Then what is printing out over serial?

Does anyone have any idea what's going on here?

Thanks!

My code and info:

MicroPython v1.19.1 on 2022-06-18; Raspberry Pi Pico with RP2040
Thonny 3.3.13

```
def main_code():
    
    while True:

        if not btn.value():                        # If button is pressed
            
            time.sleep_ms(1500)                    # It has to be held down 1500ms
            
            if not btn.value():                    # Start closing procedure
                close()


def close():
    
    motor_fwd()                                    # Start motor until both endstops trigger
        
    while endstop_1.value() or endstop_2.value():
        
        print(endstop_1.value(), endstop_2.value(), time.ticks_ms())
        #time.sleep_ms(20)
        led.value(1)
        
    motor_stp()
    time.sleep_ms(500)
    
    motor_rev()
    time.sleep_ms(5000)                            # motor has internal endstop
    
    motor_stp()
    
    led.value(0)
        
        
def motor_fwd():
    me.value(1)
    m1.value(1)
    m2.value(0)
def motor_rev():
    me.value(1)
    m1.value(0)
    m2.value(1)
def motor_stp():
    me.value(0)
    m1.value(0)
    m2.value(0)

# ===================================
#    _____ ______________  ______ 
#   / ___// ____/_  __/ / / / __ \
#   \__ \/ __/   / / / / / / /_/ /
#  ___/ / /___  / / / /_/ / ____/ 
# /____/_____/ /_/  \____/_/


from machine import Pin
import time

led = Pin(25, Pin.OUT)
led.value(0)

btn = Pin(15, machine.Pin.IN, machine.Pin.PULL_UP)

endstop_1 = Pin(8, machine.Pin.IN, machine.Pin.PULL_UP)
endstop_2 = Pin(9, machine.Pin.IN, machine.Pin.PULL_UP)

me = Pin(12, Pin.OUT)                                     # motor enable
me.value(0)
m1 = Pin(10, Pin.OUT)                                     # motor direction pin 1
m1.value(0)
m2 = Pin(11, Pin.OUT)                                     # motor direction pin 2
m2.value(0)


if __name__ == '__main__':
    
    main_code()
```

**Comment by robert-hh** [comment]
- Is it the same ticks_ms value which is printed? or does it really print the text `ticks_ms` which is printed?
- How do you run the code. Do you copy it to the board to a python file and start it from REPL, or do you use Thonny's run command? If the latter, what happens if you upload the file to the board and start it manually, may w/o using Thonny.

**Reply by MrTinkerman** [reply]
Thanks,

What keeps getting printed is: 
(triggering one of the end stop low, does not print 0 1 8014207)
```
1 1 8014207
1 1 8014208
1 1 8014208
1 1 8014208
1 1 8014208
1 1 8014208
1 1 8014208
1 1 8014208
1 1 8014208
1 1 8014209
1 1 8014209
1 1 8014209
1 1 8014209
1 1 8014209
1 1 8014209
1 1 8014209
1 1 8014209
1 1 8014209
1 1 8014210
1 1 8014210
1 1 8014210
1 1 8014210
1 1 8014210
1 1 8014210
1 1 8014210
1 1 8014210
1 1 8014210
1 1 8014211
```
The ticks_ms keep increasing.

Based on your second suggestion I tried this:

-- Save main.py to the board with this loop:
```
    while endstop_1.value() or endstop_2.value():
        
        print(endstop_1.value(), endstop_2.value(), time.ticks_ms())
        led.value(1)
```
-- In Thonny: Tools > Open system shell > CTRL-D : 'MPY: soft reboot'

Now, when the loop runs the values for the end stop are updated and the printout stops when both end stops have been triggered.
And that was what I was expecting to happen.

If I use the run command in Thonny again, it's back to as described in the opening post. 

It's not a real problem since the board controls the motor as expected but I'm still curious as to why the print statement keeps running.

Thanks

**Reply by robert-hh** [reply]
Thonny magic?

**Reply by MrTinkerman** [reply]
Yeah, let's call it that, Thonny magic :)

I do like it's features and simplicity but sometimes.... it can be buggy.
